Check markdown rs

Christopher-C-Robinson opened the pull request:
Draft of markdown parser

Christopher-C-Robinson committed:
added file

Signed-off-by: Domesticcadiz <christopher.cadiz.robinson@gmail.com>

Christopher-C-Robinson committed:
skeleton files

Signed-off-by: Domesticcadiz <christopher.cadiz.robinson@gmail.com>

Toreseen committed:
Commiting validate_document_structure.rs. Also testing to see if Christ can see this commit!

Signed-off-by: Max Wendlandt <101139274+Toreseen@users.noreply.github.com>

Christopher-C-Robinson committed:
latest code

Signed-off-by: Domesticcadiz <christopher.cadiz.robinson@gmail.com>

Christopher-C-Robinson committed:
removed binary files

Signed-off-by: Domesticcadiz <christopher.cadiz.robinson@gmail.com>

jodh-intel commented:
Speaking to @Christopher-C-Robinson and @Toreseen, we know that the following needs to be done to this draft PR:

- [ ] Remove the `target/` binaries introduced by the editor we believe.
- [ ] Remove the `src/main_from_go.rs` and `src/validate_structure.rs`.
- [ ] Possibly split the functionality out of `src/main.rs` into one or more other files below `src/`.
- [ ] Squash all the commits down into one.
- [ ] Update the commit format to match https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#patch-format.

jodh-intel commented on cmd/check-markdown-rs/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+    // Match the node value to handle different node types
+    match &node.data.borrow().value {
+        // If the node is a Heading node, print its level and text
+        NodeValue::Heading(heading) => println!(
+            "{}Header (level {}): {}",
+            indent,
+            heading.level,
+            get_node_text(&node)
+        ),
```
I know these blocks are small, but I'd be inclined to create a set of functions to handle each node type:

- `handle_heading_node(indent, level, text)`
- `handle_text_node(indent, text)`
- `handle_link_node(indent, title, url)`

The advantages of doing this include:

- simplifies `print_nodes()` and makes it easier to read.
- makes it easier to unit test different types of nodes raising confidence in the parts of the code you have tested formally.

Christopher-C-Robinson replied:
I think I added what you were looking for.

jodh-intel commented on cmd/check-markdown-rs/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+                if level == 1 {
+                    h1_count += 1;
+                    if h1_count > 1 {
+                        return Err(String::from("Error: More than one H1 heading found."));
+                    }
+                }
+
+                if level > last_seen_level + 1 {
+                    return Err(String::from("Error: Invalid heading level order."));
```
- Can you remove the `Error: ` prefix from the errors? The type of these objects `Result::Err` already encodes that they are errors so it's best to let the higher levels handle adding prefixes like this (see comment on that topic below).
- Can you include the heading in the error being returned here? In fact, ideally, you'd also add the line number and the filename to the error (either here or at a higher level) to make it 100% clear to the user which part of which file contains the error. This is particularly important when the tool starts to jump between files when it finds a cross-file link, and then finds, for example, an invalid link in another file _with the same name as a link in the initial file_ (yep - I've been there and done that ;)

Toreseen replied:
This should be fixed now!

jodh-intel commented on cmd/check-markdown-rs/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+    match link_validation_result {
+        Ok(_) => println!("  All links are valid."),
+        Err(errors) => {
+            println!("  Found {} invalid links:", errors.len());
+            for error in errors {
+                println!("    {}", error);
+            }
```
I like the way you've collected all the errors together here and are displaying them at the end. Conventionally, errors should be displayed to standard error stream, so can you change the `println!()` calls to `eprintln!()` and add an `"ERROR: "` prefix to the call in the for loop?

jodh-intel reviewed (commented):
Thanks @Christopher-C-Robinson and @Toreseen - this is looking good! I've added a few comments.

More generally, I know we've discussed this. But looking at the code I do think it might make sense to split the main into a few other files as this grows. Maybe:

```
src/output.rs
src/node.rs
src/toc.rs
src/stats.rs
```

... and then add a corresponding `mod output`, `mod node`, _etc_ in `src/main.rs` to "activate" these modules.

Also, it would be good to add a few [unit tests](https://github.com/kata-containers/kata-containers/blob/main/docs/Unit-Test-Advice.md#rust-table-driven-tests) as that will help shake out bugs and increase your confidence in the functions as you'll know they can tolerate all types of nasty input (writing tests can be fun as you try to find lots of devious inputs for your functions to handle :smiling_imp:)

Toreseen committed:
Commiting small changes for Chris to get on his laptop

Signed-off-by: Max Wendlandt <101139274+Toreseen@users.noreply.github.com>

Toreseen committed:
Delete main_from_go.rs

Unneeded file

Toreseen committed:
Delete validate_structure.rs

Unneeded file

Toreseen committed:
Commit attempt number 14. Lets see how this goes.

Signed-off-by: Max Wendlandt <101139274+Toreseen@users.noreply.github.com>

Toreseen committed:
Commit number 2, sort of another test to see if this works

Signed-off-by: Max Wendlandt <101139274+Toreseen@users.noreply.github.com>

Toreseen committed:
Delete .rustc_info.json

Toreseen committed:
Delete CACHEDIR.TAG

Toreseen committed:
Removed the ERROR prefix from the error  messages.

Signed-off-by: Maxwell Wendlandt <maxjman101@gmail.com>

Christopher-C-Robinson committed:
updated readme files

Signed-off-by: Domesticcadiz <christopher.cadiz.robinson@gmail.com>

Toreseen commented on cmd/check-markdown-rs/src/main.rs:
```diff
@@ -0,0 +1,318 @@
+        eprintln!("Usage: {} <input_file>", args[0]);
+    }
+
+    let input_file_path = env::current_dir()?.join("src").join(args[1].clone());
```
Im not sure if this is what you are lookin for, bit I'm pushing a commit now that removes the SRC from the path, assuming that you will already cd into the src directory.

Toreseen committed:
Fixed SRC being using in input_file_path.

Signed-off-by: Maxwell Wendlandt <maxjman101@gmail.com>

Christopher-C-Robinson committed:
updated readme

jodh-intel commented on cmd/check-markdown-rs/src/main.rs:
```diff
@@ -0,0 +1,317 @@
+    let arena = Arena::new();
+    let args: Vec<String> = env::args().collect();
+
+    if args.len() < 2 {
+        eprintln!("Usage: {} <input_file>", args[0]);
+    }
```
- You need to call ` std::process::exit(1)` after displaying the message: currently, the message is displayed but the program keeps running, then fails:
   ```bash
   $ check-markdown-rs
   Usage: check-markdown-rs <input_file>
   thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1', src/main.rs:22:52
   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
   ```
- Please can you add some argument parsing code so that the user can run:
  ```bash
  $ check-markdown-rs --help
  $ check-markdown-rs --version
  ```

Toreseen committed:
Added line number's to be output in error messages where the error occurred in the markdown doc

Signed-off-by: Maxwell Wendlandt <maxjman101@gmail.com>

jodh-intel reviewed (commented):
Hi @Christopher-C-Robinson and @Toreseen,

As we're very close to the end of the outreach project now, I suggest you rework this PR to remove the binary files, _etc_ as follows. Note that once you've done this though, @Toreseen won't be able to `git pull` from @Christopher-C-Robinson's fork, so you'll both have to work together and have just @Christopher-C-Robinson `git push -f` to his fork once you've followed the steps below. The steps below are similar in principle to what we did in the meeting on Tuesday with Alec and Jenny's branch. Yes, this is "hacky" in the sense that we're not using `git` to resolve the problems - we're simply recreating the branch, but I think this approach is simpler to understand. As always, make sure you have a copy of your local branch/repo before doing the following:

```bash
# Update your copy of the main branch
git checkout main
git pull

# Switch to your feature branch
git checkout check-markdown-rs

# Save the files you've created for this new tool
cp cmd/check-markdown-rs/Cargo.toml cmd/check-markdown-rs/README.md cmd/check-markdown-rs/src/main.rs /tmp/

# Rename the original version of the branch (it's now a "backup")
git branch -m check-markdown-rs-original

# Create a new clean version of the feature branch
git checkout main
git checkout -b check-markdown-rs

# Recreate the branch
mkdir cmd/check-markdown-rs
cd cmd/check-markdown-rs
cp /tmp/Cargo.toml /tmp/README.md .
mkdir src
cp /tmp/main.rs src

# Add the new files to git
git add Cargo.toml README.md src/main.rs

# Save the changes
git commit -as

# Update your fork (which will automatically update this PR)
git push -f
```

Once you've done that, you'll also need to update `cmd/README.md` to add an entry for your new tool, so:

- Add an entry for your new tool to `cmd/README.md`
- `git commit -as`
- Merge the new commit with the main commit by running `git rebase -i main`.
- Change the 2nd commit to `squash`.
- Save and quit the editor.
- Update the PR again by running, `git push -f`

jodh-intel commented on cmd/README.md:
```diff
@@ -1,13 +0,0 @@
-# CLI tools
```
The instructions above show you how to recreate the entire branch which will resolve this issue :smile:

Christopher-C-Robinson replied:
I see that now. I added that back in. should be pushed sometime tonight.

Christopher-C-Robinson committed:
updated to validate links

Christopher-C-Robinson commented:
@jodh-intel 

@Toreseen and I have done some serious work tonight on our code. If you get a chance, can you take a look at it. Might be messy bit it works. We both spent about 12 hours each on it individually. I’m finally off to bed. Hopefully we are getting close!

Toreseen committed:
Fixed one small indentatio error

Signed-off-by: Maxwell Wendlandt <maxjman101@gmail.com>

Christopher-C-Robinson committed:
added helper functions. fixed deleted readme.md

Toreseen committed:
Added heading titles and file names to the document heaing validation function.

Signed-off-by: Maxwell Wendlandt <maxjman101@gmail.com>

jodh-intel commented on cmd/check-markdown-rs/src/main.rs:
```diff
@@ -0,0 +1,568 @@
+    let arena = Arena::new();
+
+    if args.len() < 2 {
+        eprintln!("Usage: {} <input_file>", args[0]);
+        return Ok(());
```
Nit: Ideally, this would return an `Err` as this is it invalid to call the tool without specifying a file.

Christopher-C-Robinson replied:
i think this is fixed now

jodh-intel reviewed (commented):
Thanks @Christopher-C-Robinson and @Toreseen ! This is very close now.

Before we can get this landed though, you'll need to clean up the branch to (a) remove all the binary files; and (b) collapse all your commits into a single one - see my big "branch recreation" comment.

Christopher-C-Robinson committed:
added comments. fixed one of james's comments

Christopher-C-Robinson committed:
split into multiple files

Christopher-C-Robinson commented:
> Thanks @Christopher-C-Robinson and @Toreseen - this is looking good! I've added a few comments.
> 
> More generally, I know we've discussed this. But looking at the code I do think it might make sense to split the main into a few other files as this grows. Maybe:
> 
> ```
> src/output.rs
> src/node.rs
> src/toc.rs
> src/stats.rs
> ```
> 
> ... and then add a corresponding `mod output`, `mod node`, _etc_ in `src/main.rs` to "activate" these modules.
> 
> Also, it would be good to add a few [unit tests](https://github.com/kata-containers/kata-containers/blob/main/docs/Unit-Test-Advice.md#rust-table-driven-tests) as that will help shake out bugs and increase your confidence in the functions as you'll know they can tolerate all types of nasty input (writing tests can be fun as you try to find lots of devious inputs for your functions to handle 😈)

@jodh-intel , I split the main.rs into different files. let me know what your thoughts are. I will work more on it tonight.

Christopher-C-Robinson commented on cmd/check-markdown-rs/README.md:
````diff
@@ -0,0 +1,57 @@
+## Usage
+
+```
+cargo build --release
+./target/release/markdown_parser input_file.md
````
sounds good!

Christopher-C-Robinson committed:
install command avoid target line being incorrect

Toreseen committed:
Added copyright (c) to rust  files

Signed-off-by: Maxwell Wendlandt <maxjman101@gmail.com>

Toreseen committed:
Added the functionaity for H1 heading errors to ouput line number + other requested freatures.

Signed-off-by: Maxwell Wendlandt <maxjman101@gmail.com>

jodh-intel reviewed (commented):
Thanks @Christopher-C-Robinson and @Toreseen - this is looking good.

Could you remove the binary files from this PR now? Also, could you delete the unused functions that `cargo build` lists?

Christopher-C-Robinson committed:
cleaned up unused methads and variables
